### ISSUE TITLE
Opening a workspace with a cached build

### ISSUE THREAD

**Issue by Cynical-Optimist** [buildstream-migration/bst-staging#311]
[See original issue on GitLab](https://gitlab.com/BuildStream/buildstream/-/issues/311)
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Mar 21, 2018, 14:04

depends on #21

We need a way for a workspace to use a cached build instead of building a fresh one from sources.
E.G.

One could, for instance, desire to hack on WebKit and create a workspace using the last known build of the correct cache key, so after modifying some file, only part of the webkit build would have to run.

### Override workspace if cached build found

At the point of opening a new workspace, check for an appropriate cached build tree created by that element and stage that in place of the elements sources.

If no cache found, continue as normal

Add the flag `--no-cache`: This will bypass the cache and create a workspace as normal

We should have some STATUS messages along the way informing the user what is being used.

#### Workspace special sauce

> When opening workspaces, the Source API has an opportunity to enhance things. 
> 
> This is mostly because we want to be sure that say, for a git source, the upstream origin is set to the real upstream, and not set to the local artifact cache from whence it was staged.

> For right now, it would be best that you implement the feature while bypassing init_workspace() completely, as a first step

> After that, we have to come up with a plan to address it, depending on how complex that plan is; we may merge the initial implementation "without the special init_workspace() sauce" and punt that part as an enhancement

> Or, we may block the patch on first fixing that so that it works nicely off the bat

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Mar 21, 2018, 14:04

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Mar 21, 2018, 14:35

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Mar 21, 2018, 14:37

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Mar 21, 2018, 14:56

We should use the cached build tree by default, this is most convenient for the user I think.

> Always use cached builds unless a flag is used to stage the sources (bst checkout --clean?)

* This is `bst workspace open`, not `bst checkout`
* I think an option `--no-cache` could be an appropriate name

> Same as the previous option, with the addition that, if no cached build is found, it will stage the sources of the element as normal.

This is undesirably complex I think, things should "just work" without annoying errors that there was no cached artifact for the element at hand.

Instead, as a first iteration: Just use the artifact if it exists

If it becomes interesting, desirable later: We can later add an option to explicitly ignore the artifact if it exists.

Under no circumstances would we obnoxiously *fail*, just because we *assume* that the user is dead set on using a cached artifact.

Instead, we should have some STATUS messages along the way informing the user what is being used.

## Workspace special sauce

When opening workspaces, the Source API has an opportunity to enhance things. This is mostly because we want to be sure that say, for a git source, the upstream origin is set to the real upstream, and not set to the local artifact cache from whence it was staged.

How do you intend to address this part ?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Mar 23, 2018, 11:18

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Mar 27, 2018, 09:43

mentioned in commit 66e2559955d112f53c311a91e11cb95fdc521f65

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Mar 27, 2018, 15:28

mentioned in commit 2ada6f4e4ce6979800678170120a1125d7886c70

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 4, 2018, 09:29

mentioned in commit dd782ba9abd96e162307be72663e4487246c205f

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 9, 2018, 12:34

mentioned in commit 5da352be281cc1da77697e70426e65483dade0a0

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 9, 2018, 15:57

created branch [`311-opening-a-workspace-with-a-cached-build`](https://gitlab.com/BuildStream/buildstream/compare/master...311-opening-a-workspace-with-a-cached-build)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 9, 2018, 16:01

mentioned in merge request !370

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 9, 2018, 16:01

mentioned in merge request !371

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 9, 2018, 16:02

mentioned in merge request !372

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 9, 2018, 16:05

mentioned in commit c4d89ca3898f49481bc78b72cbc2408c8325b58e

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 9, 2018, 16:17

mentioned in commit 3468e365ada3e389f99d8cd8f10b89f7cc4fc58d

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 9, 2018, 16:24

mentioned in commit 99d78bd26a78137e45a98def960e7d820e623b55

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 9, 2018, 16:39

mentioned in commit d55df4805f2458d19d3a7aee24854d0e6cebd38f

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 11, 2018, 17:36

mentioned in commit 37e95a7d30de075c18c440b1d55e9ff2092743af

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 13, 2018, 10:25

mentioned in commit 272c171f3f8970224ba7595f5e17e2f8e65a647e

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 13, 2018, 11:46

mentioned in commit 545f819981e786ded74364223f99f599333993f3

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 13, 2018, 11:53

mentioned in commit cb45fd0acb9e717b80a23b8ea9cd6208e7b9cc51

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 13, 2018, 12:34

mentioned in commit 69f51b0d688c7bf6a8c74462246d7999668feb2c

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 13, 2018, 12:43

mentioned in commit 8585982af206913b949885c7dcf5e5dd8f366e1c

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 13, 2018, 12:56

mentioned in commit 3425346768b164ec07639e781fb0140b2f483880

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 13, 2018, 15:40

mentioned in commit 9d610841fbae9e62f60600acd41c6ebfffc7a581

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 13, 2018, 17:37

mentioned in commit 190c2ad9133bb55d2cdb4d1ea584f7be3def70fa

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 16, 2018, 12:58

[[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) re "Workspace special sauce"

For now i'm writing it to bypass any additional functionality.
Once I have a working, mergeable version, we can discuss what can/should be carried across when using a cached build tree.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 19, 2018, 11:36

mentioned in commit 1ecc46756e47ca2973e4780ea2c2fb5b97bdb0bd

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 19, 2018, 12:25

mentioned in commit e17f8fd30f60793ced78f3f735fa84a8d566fc75

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 19, 2018, 13:31

Ok, so there are a few things preventing this from landing.


## Caching build trees is too big

For some repositores, e.g. WebKit, we find the `.git` directory is around 6GB in size: https://gitlab.com/BuildStream/buildstream/issues/294#note_68634241

Which is rather small, compared to the additional ~20GB of "stuff" lying around in a typical WebKit developer's workspace directory.

* This should depend on a separate step which ensures that we avoid staging VCS related data into the sandbox, under every circumstance.
  * This means we have to consider that a few modules *actually depend on having the .git/* repository around at build time. I hope that we can successfully ignore this detail and say that it's acceptable to have to work around this very weird assumption.
* This also means that we should probably not, under any circumstance, ever be caching a workspace build directory
  * It would first of all contradict this particular issue title, you cannot checkout a cached build directory for a workspace of the same key
  * But caching of build directories serves more purposes than just opening workspaces, these purposes would have to be disabled when a workspace was used to generate the artifact

## Workspace uses `Source.init_workspace()`

This is rather less pressing but worth mentioning.

The semantics of `Source.init_workspace()` is to *replace* staging, not *augment* it. This means that without additional work, you would not get the nice experience of having your workspace directory for a git checkout have the correctly assigned *origin*


## Some thoughts

So far we have some thoughts for a solution.

This needs some more thought but, essentially we should disable staging of VCS data at `Source.stage()` time in general, this should prevent the build directory from getting too big.

In the case that you need to open a workspace on a built artifact, we could employ a "fill in the gaps" technique:
* First stage the cached build directory into the target workspace
* Next stage the actual `Source` *including* the VCS data, *with* the origin redirection to the proper upstream URL, into a temp directory
* Copy files which exist in the temp directory but not in the target workspace, over into the workspace (fill in the blanks)

This will necessarily involve some aggressive Source API refactoring, I think it's acceptable to deprecate and even remove `Source.init_workspace()` as I think the likelyhood of it being used anywhere outside of BuildStream is very slim to none.

The `Source.stage()` API on the other hand needs to stay in tact, but it could be made to additionally do what Source.init_workspace() was doing *automatically*, and we may have to add a separate codepath `Source.stage_lean()` which would do the mumbo jumbo without including any VCS data. The core could catch an `ImplError` should `Source.stage_lean()` not be implemented and fallback to regular `Source.stage()` in that case (in fact, `Source.stage_lean()` would only be necessary in Source plugins which behave like `git` and place VCS data into your checkout).

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Apr 23, 2018, 13:17

Some follow up:

* This now is blocking on #376 
  * This should make it reasonable to cache build trees at all
* Workspace build output should never get cached, period.
* When checking out a workspace on a cached build:
  * First checkout the build sandbox
  * Next use Source.init_workspace() to checkout the data into a temp directory
  * Now overwrite anything in the build sandbox checkout with the workspace checkout

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Apr 24, 2018, 11:59

[[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) 
>  Now overwrite anything in the build sandbox checkout with the workspace checkout

Do you mean "overwrite build sandbox with cached build?
just confirming

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 1, 2018, 12:11

mentioned in commit 172c3706b4983e7493315937c8fd0c740bd1e49f

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 2, 2018, 09:59

changed the description

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 4, 2018, 15:08

mentioned in commit daaa8d97adc30c20c1f57f131b8ec29bfb6f4f87

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 10, 2018, 09:03

mentioned in commit e3cf8087c3aa4b1f4f8db1a19eba77fb77db91ed

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 10, 2018, 09:03

mentioned in commit d5d3b09ea2d60546abc9ec4cffdbbf9d48d28219

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 10, 2018, 15:13

mentioned in commit bd113a12517c9f3fc8af0a7d8d96e22ab5a4115a

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 10, 2018, 15:13

mentioned in commit db7c916f80c1a5fe49dba0ec9df6ced5a02b0b86

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 23, 2018, 11:40

assigned to [[Gitlab user @knownexus]](https://gitlab.com/knownexus)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on May 24, 2018, 12:49

marked this issue as related to #21

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on May 24, 2018, 12:49

marked this issue as related to #376

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on May 25, 2018, 11:30

mentioned in merge request !474

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @toscalix]](https://gitlab.com/toscalix) on May 31, 2018, 14:21

!472 replaced by !474

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Jun 5, 2018, 20:45

mentioned in merge request !475

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @toscalix]](https://gitlab.com/toscalix) on Jun 24, 2018, 15:32

[[Gitlab user @knownexus]](https://gitlab.com/knownexus) what is the current status of this issue? Should we send it back to ~"To Do" ?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Jul 4, 2018, 18:00

I am moving back to ToDo since:
* I know that it currently not being worked on.
* [[Gitlab user @toscalix]](https://gitlab.com/toscalix) received no response
* I'm trying to get the gitlab WIP items into good shape before GUADEC

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Jul 4, 2018, 18:01

unassigned [[Gitlab user @knownexus]](https://gitlab.com/knownexus)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @albfan]](https://gitlab.com/albfan) on Jul 9, 2018, 10:06

Maybe related? https://gitlab.com/BuildStream/buildstream/issues/21#note_84054783

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Jul 27, 2018, 12:03

assigned to [[Gitlab user @knownexus]](https://gitlab.com/knownexus)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Aug 6, 2018, 15:44

unassigned [[Gitlab user @knownexus]](https://gitlab.com/knownexus)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @knownexus]](https://gitlab.com/knownexus) on Aug 6, 2018, 15:46

Whoever this ends up being assigned to, if you need any assistance just give me a ping,
anything i could put on here would likely be outdated and wrong by the time it comes to being used.

That said, my latest attempt was on the branch: caching_build_trees : 8a45fc8e5b6acbbcc5cede80542bdfb2438fe827

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @sstriker]](https://gitlab.com/sstriker) on Aug 9, 2018, 10:46

> Add the flag `--no-cache`: This will bypass the cache and create a workspace as normal

Minor detail, if there is a flag, there should also be a configuration option in my opinion.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @bethwhite]](https://gitlab.com/bethwhite) on Sep 3, 2018, 13:50

Addressed in the workspaces proposal on the mailing list [here][1]

[1]: https://mail.gnome.org/archives/buildstream-list/2018-August/msg00078.html

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Sep 28, 2018, 13:41

This was included in the workspace proposal summary: https://mail.gnome.org/archives/buildstream-list/2018-September/msg00036.html

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Oct 12, 2018, 16:34

mentioned in merge request !873

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Oct 25, 2018, 13:44

assigned to [[Gitlab user @tpollard]](https://gitlab.com/tpollard)

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Oct 29, 2018, 17:18

Hi [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) just wanted to get some input on what I've currently been doing. In !873 I'm trying to implement the naive functionality that has been discussed on this issue. I think there's some cleverer stuff we can do on a specific source plugin basis (specifically with gits implementation of init_workspace_), but for now I've gone with the brute force method of:

- Stage the source into the workspace temp dir as per usual
- Checkout the cached buildtree artifact (this is currently based off of !786 for buildtree interaction) into another  tempdir dir
- Overwrite any file in the source workspace temp dir with the buildtree cached equiv

To do this I'm breaking out a call to the host systems 'cp', as I've not found shutil.copytree or utils.py copy_files (both of these can't handle directories in which files exist with the same name), shutil.move (will place a directory inside another if it already exits) or distutils.file_util.copy_file (can't handle non regular files) appropriate for this specific task. I don't really like using the host cp (as it probably has slight behaviour changes in OSX for example, and obviously is not in Windows at all) but I don't know enough about the intricacies of the file operations to know if utils.py can be used here.

As it stands I'm using the cp flags -pfr, with 'p' preserving the mode, ownership and timestamps. Taking ostree as an example element, this leads to the workspace .git picking up a change in mode for grub file. I don't know if this is perceived behaviour that was envisioned for this issue.

Any feedback on this comment is appreciated, thanks.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Oct 30, 2018, 09:41

Hi [[Gitlab user @tpollard]](https://gitlab.com/tpollard) 

This probably calls for an enhancement of the `utils.copy_files()` APIs, calling host `cp` is certainly not the way forward here.

Regarding file modes, these are supposed to be ensured to be deterministic at stage time, such that files are staged with the same mode bits regardless of host operating system (or notably, the user's `umask` setting), a series of patches was landed by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) this year to ensure this determinism.

It is possible that some non-determinism remains when staging workspaces to the build sandbox's source directory; or when staging sources to the workspace. We should verify that we have measures in place to ensure determinism of the mode bits in these areas too.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Oct 30, 2018, 11:48

Thanks for the input [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb), I definitely think extending in utils to handle this scenario is the right way to go, using cp in the WIP however is at least useful for progessing.

[[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) I wonder if you could provide me with any pointers around the expected file behaviour of overwriting source files in the workspace with those from the cached buildtree? My example of a grub file that had it's mode modified during the build as an example, what exact file data/metadata should/should not be preserving from the buildtree file state. I get the point of being deterministic in the sense of avoiding any host configuration when stagining sources for being built, but this scenario is an extension from that

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Oct 30, 2018, 13:35

mentioned in merge request !869

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Oct 30, 2018, 20:28

[[Gitlab user @tpollard]](https://gitlab.com/tpollard), [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid); one more point comes to mind.

The introduction of CAS has regressed issue #38 to a worse state than we had with ostree, this is a known issue.

What this means is that currently, whatever mode bits are stored in an artifact or in CAS is currently lost, except for the boolean distinction of whether the file was executable or not.

See [this comment](https://gitlab.com/BuildStream/buildstream/issues/38#note_87327672) for further context, I really hope we can focus on fixing this in CAS soon but I am doubtful that we can fix it for 1.4.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on Nov 1, 2018, 10:12

Are not the permissions on the cached built files deterministic anyway?

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @LaurenceUrhegyi]](https://gitlab.com/LaurenceUrhegyi) on Nov 5, 2018, 09:30

marked this issue as related to #740

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @jonathanmaw]](https://gitlab.com/jonathanmaw) on Nov 5, 2018, 11:01

mentioned in merge request !786

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Nov 7, 2018, 11:02

Hi [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) sorry that I've not been able to reply sooner, I've been tied up in other work. In regards to this work specifically the topic of determinism is something that I'm not too sure about, mainly because I'm not aware if it's needed for this situation. https://gitlab.com/BuildStream/buildstream/issues/311#note_112907326 has my current progress, and the rest of this ticket has the background information (the WIP !873 too). I'm achieve the generic goal of overlaying the buildtree over the source workspace, but the specifics of that action are what I'm looking for direction with.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on Nov 13, 2018, 11:38

After talking [[Gitlab user @tpollard]](https://gitlab.com/tpollard) today this is my opinion on this issue.

I do not think determinism is a big issue. If your function copying files is deterministic, then the result is deterministic.

You could want to have determinism between non workspaced and workspaced build. In my opinion, this should not be that important. If we add a `.git` directory, we lost determinism anyway. And cache keys are different.

We sure want to be close to the original build however. But we will be limited at the end by how much information is stored in the build artifact. What we need to make sure is that the functions copying files can copy as much information that checking out build artifact.

Also `utils.copy_files/utils.link_files` transform absolute symbolic links to relative. Is this something we want for build artifacts? There is something specific to `utils.copy_files/utils.link_files` and maybe we should consider not using them everywhere.

I want to come back to why we need to copy the files here. It is to be able to overlay the build artifact checkout on top of the staged source.

`Source.init_workspace` does not "augment" as [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) said. But we need something that augments. Here is why.

If the build system removes a source file, it will appear in the workspace. You might wonder why we should care about builds that remove source files. But that might be a bug that the use is investigating. What surprise it would be for the user to open the workspace with the build and see that the file is still there.

But the most import issue has to do with the shallow clone. The build will optionally contain a shallow clone. Only the source plugin can deal with this and know the full clone has priority. So this means that plugins need to be called after the build artifact is checked out. And then it needs to "augment" the repository. This should be done by a new optional method to Source.

And in this case, we do not need to worry about copying files.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Nov 20, 2018, 14:00

As https://gitlab.com/BuildStream/buildstream/merge_requests/786 has now landed into master, we have the basis of the ability to inspect and further download the partial artifact subdir (i.e, just download the missing buildtree object content) on demand. To fully close the issue https://gitlab.com/BuildStream/buildstream/issues/494 it is expected/required that situations such as trying to open a workspace cached buildtree on a partial artifact would spawn a scheduler job to attempt the download.

As it stands the current first 'pass' (as mentioned already in this ticket) for workspace buildtrees is to simply dropback to existing source staging behaviour if the buildtree is not cached locally (or overridden to not use the cache via cli/config). With that said it will not close 494 in its initial state.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tpollard]](https://gitlab.com/tpollard) on Nov 21, 2018, 13:09

>  But the most import issue has to do with the shallow clone. The build will optionally contain a shallow clone. Only the source plugin can deal with this and know the full clone has priority. So this means that plugins need to be called after the build artifact is checked out. And then it needs to "augment" the repository. This should be done by a new optional method to Source.

Just checking that I'm parsing this correctly [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid), are you suggesting that the buildtree/cached build is copied into the empty dir containing the .git, and then git only has to 'augment' the differences? As it stands git init_workspace does a clone with no checkout initially, so this would provide the opportunity to stage the buildtree

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on Nov 23, 2018, 23:41

[[Gitlab user @tpollard]](https://gitlab.com/tpollard) No. I am saying that the cached build will contain *a* `.git` directory in the future. It just happens that right now, `.git` is not part of the build. This `.git` directory will not contain the full clone. It will be something shallow that is just enough for `git describe` to work (see #487). Even possibly something that does not even contain any data able to make any checkout.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) on Nov 23, 2018, 23:45

If you copy the cached build after cloning `.git` directory, then it will be overwritten and possibly corrupted.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @tristanvb]](https://gitlab.com/tristanvb) on Nov 26, 2018, 13:04

Just ringing in here with a bit of a review of what has been discussed here as [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) requested.

First, [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) is correct that, if we first initialize the workspace using the regular plugin `init_workspace` into a directory, and then stage a build tree on top of it; we risk that the cached build tree corrupts the `.git/` directory in the case of a `git` source. This is because of the long standing blocker issue #487 which is presently, and necessarily being rectified.

Second, while [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) is correct that the `utils` functions which copy/link files deterministically while preserving as much information as possible, undesirably mangle input symlinks, it is **very** undesirable to start having separate codepaths for this activity.

Further, the reason why we mangle symlinks in the first place is questionable, it is a bit paranoid I admit, and is mostly a measure to avoid mistakenly writing outside of the staging area at any time that we are not actually running code in the sandboxed environment. To sum it up; the fact that we normalize symlinks to be relative paths is itself a bug, because we fail to accurately preserve the exact content which might have been produced during a build, instead we modify the paths to be relative, which is technically wrong and should eventually be fixed.

If we can do anything to avoid employing separate codepaths for `utils.copy_files()`/`utils.link_files()`, then we should avoid them. Maybe as a first step, we can grant these functions (essentially `utils._process_list()`) the ability to do their job without symlink mangling, with an internal hidden private keyword argument or such, only known by BuildStream internals... Then we could eventually transition to not mangling symlinks at all in the future, fixing the above mentioned issue and removing the optionality of it.

Third, [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) points out that we could have a new method for plugins to implement the augmentation of a checked out build tree into a workspace. This does seem like a plausible approach, but it might result in unsupporting *"workspacing plus build tree"* scenarios for source types which do not implement this new method.

Instead of this, there *might* be a better option - for instance, the code which fixes #487 for the purpose of `git describe` necessarily adds some code to the `git` plugin at staging time, perhaps instead we could have a separate method for *undoing* the thing that the `git` plugin needs to do for the sake of `git describe`, such that we could undo it generically before generically overlaying a build tree ?

Finally, I'm not sure if the above alternative can work, because technically speaking; [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid) is also correct in his analysis that modified and removed files can be consequences of builds, in which case we would *want* the staged build tree to preserve the modifications and removals of files which left the git HEAD in a *"dirty"* state. This last bit is indeed a bit of a pickle, and maybe an argument as to why we *must* use [[Gitlab user @valentindavid]](https://gitlab.com/valentindavid)'s approach of a separate buildtree-to-workspace augmentation plugin method.

**Comment by Cynical-Optimist** [comment]
In GitLab by [[Gitlab user @willsalmon]](https://gitlab.com/willsalmon) on Dec 19, 2018, 14:52

marked this issue as related to #829
